packages/@aws-cdk/aws-iam/lib/grant.ts

```diff
@@ -0,0 +1,257 @@
+    });
+
+    if (result.success) { return result; }
+
+    const statement = new PolicyStatement()
+      .addActions(...options.actions)
+      .addResources(...(options.resourceSelfArns || options.resourceArns))
+      .addPrincipal(options.principal!);
+
+    options.resource.addToResourcePolicy(statement);
+
+    return grantResult({ addedToResource: true, statement, options });
+  }
+
+  /**
+   * Try to grant the given permissions to the given principal
+   *
+   * Absence of a principal leads to a warning, but failing to add
+   * the permissions to a present principal is not an error.
+   */
+  public static grantOnPrincipal(options: GrantOnPrincipalOptions): Grant {
+    if (!options.principal) {
+      if (options.scope) {
+        // tslint:disable-next-line:max-line-length
+        options.scope.node.addWarning(`Could not add grant for '${options.actions}' on '${options.resourceArns}' because the principal was not available. Add the permissions by hand.`);
+      }
+      return grantResult({ principalMissing: true, options });
+    }
+
+    const statement = new PolicyStatement()
+      .addActions(...options.actions)
+      .addResources(...options.resourceArns);
+
+    const addedToPrincipal = options.principal.addToPolicy(statement);
+
+    return grantResult({ addedToPrincipal, statement: addedToPrincipal ? statement : undefined, options });
+  }
+
+  /**
+   * Add a grant both on the principal and on the resource
+   *
+   * As long as any principal is given, granting on the pricipal may fail (in
+   * case of a non-identity principal), but granting on the resource will
+   * always be done.
+   *
+   * Statement will be the resource statement.
+   */
+  public static grantOnPrincipalAndResource(options: GrantOnPrincipalAndResourceOptions): Grant {
+    const result = Permissions.grantOnPrincipal({
+      ...options,
+      scope: options.resource,
+    });
+
+    if (result.principalMissing) { return result; }
+
+    const statement = new PolicyStatement()
+      .addActions(...options.actions)
+      .addResources(...(options.resourceSelfArns || options.resourceArns))
+      .addPrincipal(options.principal!);
+
+    options.resource.addToResourcePolicy(statement);
+
+    return grantResult({ addedToPrincipal: result.addedToPrincipal, addedToResource: true, statement, options });
+  }
+}
+
+/**
+ * The result of the grant() operation
+ *
+ * This class is not instantiable by consumers on purpose, so that they will be
+ * required to call the Permissions.grant() functions.
+ */
+export class Grant {
+  /**
+   * There was no principal to add the permissions to
+   */
+  public readonly principalMissing: boolean;
+
+  /**
+   * The grant was added to the principal's policy
+   */
+  public readonly addedToPrincipal: boolean;
+
+  /**
+   * The grant was added to the resource policy
+   */
+  public readonly addedToResource: boolean;
+
+  /**
+   * The policy statement that was created for this permission
```

Comment:
is that the principal's policy statement or the resource policy statement? Maybe it makes sense to supply both and use them as indications that that the specific policy was updated (instead of `addedToResource` and `addedToPrincipal`). May also result in a nicer way handle nulls without using "!".

```
principalStatement?: PolicyStatement; // undefined if the principal policy was not updated
resourceStatement?: PolicyStatement; // undefined if the resource policy was not updated
```

Reply:
Yeah sure.